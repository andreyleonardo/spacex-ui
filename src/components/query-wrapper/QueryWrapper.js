import React from 'react';
import PropTypes from 'prop-types';
import { Query } from 'react-apollo';

const QueryWrapper = ({ query, children, variables }) => (
  <Query query={query} variables={variables}>
    {({ loading, error, data }) => {
      if (loading) return <div id="loading-id">Loading...</div>;
      if (error) return <div id="error-id">Ops... Try again</div>

      return children(data);
    }}
  </Query>
);

QueryWrapper.propTypes = {
  query: PropTypes.shape({}).isRequired
};

export default QueryWrapper;
