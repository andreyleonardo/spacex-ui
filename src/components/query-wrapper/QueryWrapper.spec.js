import React from 'react';
import { mount, shallow } from 'enzyme';
import { MockedProvider } from 'react-apollo/test-utils';
import wait from 'waait';
import QueryWrapper from './QueryWrapper';
import { query, mocksFailQuery, mocksSuccessQuery } from '../../../test/utils/query-mock';

describe('Query Wrapper', () => {
  it('renders without crash', () => {
    const queryWrapper = shallow(<QueryWrapper query={{ query: 'query' }} />);
    expect(queryWrapper).toMatchSnapshot();
  });

  it('renders a loading component when its querying', () => {
    const queryWrapper = mount(
      <MockedProvider mocks={[]}>
        <QueryWrapper query={query} />
      </MockedProvider>
    );

    expect(queryWrapper.exists('div#loading-id')).toBeTruthy();
  });

  it('renders an error component when query fails', async () => {
    const queryWrapper = mount(
      <MockedProvider mocks={mocksFailQuery} addTypename={false}>
        <QueryWrapper query={query} />
      </MockedProvider>
    );

    await wait(0);

    queryWrapper.update();

    expect(queryWrapper.exists('div#error-id')).toBeTruthy();
  });

  it('renders a children component when query success', async () => {
    const queryWrapper = mount(
      <MockedProvider mocks={mocksSuccessQuery} addTypename={false}>
        <QueryWrapper query={query} >
          {({ test }) =>(
            <>
              <span id="id">{test.id}</span>
              <span id="name">{test.name}</span>
            </>
          )}
        </QueryWrapper>
      </MockedProvider>
    );

    await wait(0);

    queryWrapper.update();

    expect(queryWrapper.exists('span#id')).toBeTruthy();
    expect(queryWrapper.exists('span#name')).toBeTruthy();
    expect(queryWrapper.find('span#id').text()).toBe("1");
    expect(queryWrapper.find('span#name').text()).toBe('Test');
  });
});
