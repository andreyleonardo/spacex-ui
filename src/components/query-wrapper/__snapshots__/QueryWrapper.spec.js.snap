// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Query Wrapper renders without crash 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <QueryWrapper
    query={
      Object {
        "query": "query",
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": [Function],
      "query": Object {
        "query": "query",
      },
      "variables": undefined,
    },
    "ref": null,
    "rendered": [Function],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": [Function],
        "query": Object {
          "query": "query",
        },
        "variables": undefined,
      },
      "ref": null,
      "rendered": [Function],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
